src/node_contextify.cc

```diff
@@ -507,6 +510,208 @@ class ContextifyContext {
+
+    Maybe<bool> has = sandbox->HasOwnProperty(context, index);
+    Local<Value> descriptor_intercepted = has.IsNothing() ?
+        ctx->global_proxy()
+            ->GetOwnPropertyDescriptor(context, UIntToName(isolate, index))
+            .ToLocalChecked() :
+        sandbox
+            ->GetOwnPropertyDescriptor(context, UIntToName(isolate, index))
+            .ToLocalChecked();
+
+    // If the index had already been defined on the sandbox or global object,
+    // we intercept, query and return it.
+    // Else, there is no need to intercept with first time definitions.
+    if (!descriptor_intercepted->IsUndefined()) {
+      info.GetReturnValue().Set(descriptor_intercepted);
+    }
+  }
+
+
+  static void IndexedPropertyDefinerCallback(
+      uint32_t index,
+      const PropertyDescriptor& desc,
+      const PropertyCallbackInfo<Value>& info) {
+    ContextifyContext* ctx;
+    ASSIGN_OR_RETURN_UNWRAP(&ctx, info.Data().As<Object>());
+
+    // Still initializing
+    if (ctx->context_.IsEmpty())
+      return;
+
+    Local<Context> context = ctx->context();
+    v8::Isolate* isolate = context->GetIsolate();
+    HandleScope scope(isolate);
+
+    auto attributes = PropertyAttribute::None;
+    bool is_declared = ctx->global_proxy()
+            ->GetRealNamedPropertyAttributes(ctx->context(),
+                  UIntToName(isolate, index))
+            .To(&attributes);
+    bool read_only =
+        static_cast<int>(attributes) &
+        static_cast<int>(PropertyAttribute::ReadOnly);
+
+    // If the index is set on the global as read_only, don't change it on
+    // the global or sandbox.
+    if (is_declared && read_only)
+      return;
+
+    Local<Object> sandbox = ctx->sandbox();
+
+    auto define_prop_on_sandbox =
+        [&] (PropertyDescriptor* desc_for_sandbox) {
+          if (desc.has_enumerable()) {
+            desc_for_sandbox->set_enumerable(desc.enumerable());
+          }
+          if (desc.has_configurable()) {
+            desc_for_sandbox->set_configurable(desc.configurable());
+          }
+          // Set the index on the sandbox.
+          sandbox->DefineProperty(context,
+               UIntToName(isolate, index), *desc_for_sandbox);
+        };
+
+    if (desc.has_get() || desc.has_set()) {
+      PropertyDescriptor desc_for_sandbox(
+          desc.has_get() ? desc.get() : v8::Undefined(isolate).As<Value>(),
+          desc.has_set() ? desc.set() : v8::Undefined(isolate).As<Value>());
+
+      define_prop_on_sandbox(&desc_for_sandbox);
+    } else {
+      Local<Value> value =
+          desc.has_value() ? desc.value() :
+          v8::Undefined(isolate).As<Value>();
+
+      if (desc.has_writable()) {
+        PropertyDescriptor desc_for_sandbox(value, desc.writable());
+        define_prop_on_sandbox(&desc_for_sandbox);
+      } else {
+        PropertyDescriptor desc_for_sandbox(value);
+        define_prop_on_sandbox(&desc_for_sandbox);
+      }
+    }
+
+    printf("foo\n");
```

Comment:
Debug code again.

There's a lot of duplication.  Couldn't you implement them by calling `UIntToName(isolate, index)` and passing the result to `GlobalPropertyDefinerCallback()`, etc.?

Reply:
It's all duplication. Let me think about how to get rid of the code duplication while maintaining readability.